File: frontend/src/lib/api/patient-allergy.ts
import { api } from './client';
import type { SnomedMaster } from './loinc';

// ============================================================================
// Patient Allergy Types
// ============================================================================

export interface PatientAllergy {
  id: number;
  patient_id: number;
  visit_id?: number;
  snomed_code: string;
  snomed_display: string;
  category: AllergyCategory;
  criticality: AllergyCriticality;
  notes?: string;
  onset_date?: string;
  recorded_at: string;
  recorded_by?: number;
  satusehat_id?: string;
  satusehat_sent_at?: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  patient?: {
    id: number;
    name: string;
    medical_record_number: string;
  };
  visit?: {
    id: number;
    visit_number: string;
  };
  recorded_by_user?: {
    id: number;
    full_name: string;
  };
}

export type AllergyCategory = 'medication' | 'food' | 'environment' | 'biologic';
export type AllergyCriticality = 'low' | 'high' | 'unable-to-assess';

export interface AllergyCategoryOption {
  value: AllergyCategory;
  label: string;
}

export interface AllergyCriticalityOption {
  value: AllergyCriticality;
  label: string;
}

export interface AllergyOptions {
  categories: AllergyCategoryOption[];
  criticalities: AllergyCriticalityOption[];
}

export interface CreateAllergyInput {
  patient_id: number;
  visit_id?: number;
  snomed_code: string;
  snomed_display: string;
  category?: AllergyCategory;
  criticality?: AllergyCriticality;
  notes?: string;
}

export interface BulkCreateAllergiesInput {
  patient_id: number;
  visit_id?: number;
  allergies: {
    snomed_code: string;
    snomed_display: string;
    category?: AllergyCategory;
    criticality?: AllergyCriticality;
    notes?: string;
  }[];
}

export interface BulkCreateAllergiesResponse {
  message: string;
  created: PatientAllergy[];
  skipped: string[];
  count: number;
}

// ============================================================================
// Patient Allergy API
// ============================================================================

export const patientAllergyApi = {
  // Get allergy options (categories, criticalities)
  getOptions: () => 
    api.get<AllergyOptions>('/patient-allergies/options'),

  // Search SNOMED CT for allergy
  searchSnomed: (query: string, limit = 20) =>
    api.get<{ data: SnomedMaster[] }>('/patient-allergies/snomed/search', {
      params: { q: query, limit }
    }),

  // Get allergies by patient
  getByPatient: (patientId: number) =>
    api.get<{ data: PatientAllergy[] }>(`/patient-allergies/patient/${patientId}`),

  // Get allergy history by patient (including inactive)
  getHistoryByPatient: (patientId: number) =>
    api.get<{ data: PatientAllergy[] }>(`/patient-allergies/patient/${patientId}/history`),

  // Get active allergies count for patient
  getCountByPatient: (patientId: number) =>
    api.get<{ count: number }>(`/patient-allergies/patient/${patientId}/count`),

  // Get allergies by visit (returns all patient allergies)
  getByVisit: (visitId: number) =>
    api.get<{ data: PatientAllergy[] }>(`/patient-allergies/visit/${visitId}`),

  // Create single allergy
  create: (data: CreateAllergyInput) =>
    api.post<{ message: string; data: PatientAllergy }>('/patient-allergies', data),

  // Create multiple allergies (bulk)
  bulkCreate: (data: BulkCreateAllergiesInput) =>
    api.post<BulkCreateAllergiesResponse>('/patient-allergies/bulk', data),

  // Update allergy
  update: (id: number, data: Partial<PatientAllergy>) =>
    api.put<{ message: string; data: PatientAllergy }>(`/patient-allergies/${id}`, data),

  // Delete allergy (soft delete - sets is_active = false)
  delete: (id: number) =>
    api.delete<{ message: string }>(`/patient-allergies/${id}`),
};

// ============================================================================
// Allergy Category/Criticality Helpers
// ============================================================================

export const ALLERGY_CATEGORY_LABELS: Record<AllergyCategory, string> = {
  medication: 'Obat',
  food: 'Makanan',
  environment: 'Lingkungan',
  biologic: 'Biologis',
};

export const ALLERGY_CRITICALITY_LABELS: Record<AllergyCriticality, string> = {
  low: 'Rendah',
  high: 'Tinggi',
  'unable-to-assess': 'Tidak Dapat Dinilai',
};

export const ALLERGY_CRITICALITY_COLORS: Record<AllergyCriticality, string> = {
  low: 'bg-yellow-100 text-yellow-800 border-yellow-300',
  high: 'bg-red-100 text-red-800 border-red-300',
  'unable-to-assess': 'bg-gray-100 text-gray-800 border-gray-300',
};

File: backend/models/patient_allergy.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// PatientAllergy stores patient allergy data with SNOMED CT codes
// Used for structured allergy recording in medical records
type PatientAllergy struct {
	ID        uint           `gorm:"primaryKey" json:"id"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	// References
	PatientID uint `gorm:"not null;index" json:"patient_id"`
	VisitID   uint `gorm:"index" json:"visit_id"` // Visit where allergy was recorded

	// SNOMED CT Code (from snomed_ct table)
	SnomedCode    string `gorm:"size:20;not null" json:"snomed_code"`     // e.g., "91936005"
	SnomedDisplay string `gorm:"size:255;not null" json:"snomed_display"` // e.g., "Allergy to penicillin"

	// Category: food, medication, environment, biologic
	Category string `gorm:"size:50;not null;default:'medication'" json:"category"`

	// Criticality: low, high, unable-to-assess
	Criticality string `gorm:"size:50;default:'low'" json:"criticality"`

	// Additional info
	Notes      string     `gorm:"type:text" json:"notes"`
	OnsetDate  *time.Time `json:"onset_date,omitempty"` // When allergy first occurred
	RecordedAt time.Time  `gorm:"not null" json:"recorded_at"`
	RecordedBy *uint      `json:"recorded_by,omitempty"` // User ID who recorded (optional)

	// SatuSehat integration
	SatuSehatID     string     `gorm:"size:100" json:"satusehat_id,omitempty"`
	SatuSehatSentAt *time.Time `json:"satusehat_sent_at,omitempty"`

	// Status
	IsActive bool `gorm:"default:true" json:"is_active"` // False if allergy resolved

	// Relations
	Patient        *Patient `gorm:"foreignKey:PatientID" json:"patient,omitempty"`
	Visit          *Visit   `gorm:"foreignKey:VisitID" json:"visit,omitempty"`
	RecordedByUser *User    `gorm:"foreignKey:RecordedBy" json:"recorded_by_user,omitempty"`
}

func (PatientAllergy) TableName() string {
	return "patient_allergies"
}

// AllergyCategory constants
const (
	AllergyCategoryFood        = "food"
	AllergyCategoryMedication  = "medication"
	AllergyCategoryEnvironment = "environment"
	AllergyCategoryBiologic    = "biologic"
)

// AllergyCriticality constants
const (
	AllergyCriticalityLow            = "low"
	AllergyCriticalityHigh           = "high"
	AllergyCriticalityUnableToAssess = "unable-to-assess"
)

// GetAllergyCategoryOptions returns available allergy categories
func GetAllergyCategoryOptions() []map[string]string {
	return []map[string]string{
		{"value": AllergyCategoryMedication, "label": "Obat"},
		{"value": AllergyCategoryFood, "label": "Makanan"},
		{"value": AllergyCategoryEnvironment, "label": "Lingkungan"},
		{"value": AllergyCategoryBiologic, "label": "Biologis"},
	}
}

// GetAllergyCriticalityOptions returns available criticality levels
func GetAllergyCriticalityOptions() []map[string]string {
	return []map[string]string{
		{"value": AllergyCriticalityLow, "label": "Rendah"},
		{"value": AllergyCriticalityHigh, "label": "Tinggi"},
		{"value": AllergyCriticalityUnableToAssess, "label": "Tidak Dapat Dinilai"},
	}
}
